### PR TITLE
Improve database error management

### PR DESCRIPTION
**Pull request by adriantpaez**
## Changes:
- Change database panics for return errors

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [ ] Yes
- [x] No

**In case you checked yes, did you write tests??**

- [ ] Yes
- [ ] No

### CODE REVIEW

File: internal/db/abi/manager.go [hunk 1]
```diff
@@ -26,45 +17,35 @@ func NewABIManager(database db.Database) *Manager {
...
 	// Build the key from contract address
 	key := []byte(contractAddress)
 	value, err := proto.Marshal(abi)
 	if err != nil {
-		// notest
-		panic(any(fmt.Errorf("%w: %s", MarshalError, err)))
+		return err
 	}
 	err = m.database.Put(key, value)
 	if err != nil {
```
Comment (tshakalekholoane) [inline internal/db/abi/manager.go:45]: I suggest using the `if err = ... ; err != nil {}` idiom here.

File: internal/db/block/block_test.go [hunk 2]
```diff
@@ -41,18 +41,29 @@ func TestManager(t *testing.T) {
...
-		returnedBlock = manager.GetBlockByNumber(block.BlockNumber)
-		if returnedBlock == nil {
+		returnedBlock, err = manager.GetBlockByNumber(block.BlockNumber)
+		if err != nil {
+			if db.IsNotFound(err) {
+				t.Errorf("unexpected nil after search for block with number %d", block.BlockNumber)
+			} else {
+				t.Error(err)
+			}
 			t.Errorf("unexpected nil after search for block with number %d", block.BlockNumber)
```
Comment (tshakalekholoane) [inline internal/db/block/block_test.go:67]: This line is never reached. Typo? Also, the `else` clause can be omitted.

File: internal/db/block/manager.go [hunk 3]
```diff
@@ -20,82 +20,68 @@ func NewManager(database db.Database) *Manager {
...
 	numberKey := buildNumberKey(block.BlockNumber)
 	// Encode the block as []byte
 	rawValue, err := marshalBlock(block)
 	if err != nil {
-		panic(any(err))
+		return err
 	}
 	// TODO: Use transaction?
 	// Save (hashKey, block)
 	err = manager.database.Put(hashKey, rawValue)
```
Comment (tshakalekholoane) [inline internal/db/block/manager.go:75]: The only return value is an error so again I suggest the `if err = ...` idiom here 😅.

File: internal/db/state/storage.go [hunk 4]
```diff
@@ -20,33 +18,29 @@ func (s *Storage) Update(other *Storage) {
...
-	}
-	// Check not found
-	if rawData == nil {
-		// notest
-		return nil
+		return nil, err
 	}
 	value := new(Storage)
 	err = proto.Unmarshal(rawData, value)
 	if err != nil {
```
Comment (tshakalekholoane) [inline internal/db/state/storage.go:28]: Again, I suggest using the if err = ... ; err != nil {} idiom here.

File: internal/db/transaction/manager.go [hunk 5]
```diff
@@ -25,75 +24,59 @@ func NewManager(txDb, receiptDb db.Database) *Manager {
...
-func (m *Manager) PutTransaction(txHash types.TransactionHash, tx types.IsTransaction) {
+func (m *Manager) PutTransaction(txHash types.TransactionHash, tx types.IsTransaction) error {
 	rawData, err := marshalTransaction(tx)
 	if err != nil {
-		// notest
-		log.Default.With("error", err).Panic("error marshalling Transaction")
+		return err
 	}
 	err = m.txDb.Put(txHash.Bytes(), rawData)
 	if err != nil {
```
Comment (tshakalekholoane) [inline internal/db/transaction/manager.go:33]: Here too, I suggest using the if err = ... ; err != nil {} idiom.

File: internal/db/abi/manager_test.go [hunk 6]
```diff
@@ -18,10 +18,16 @@ func TestManager(t *testing.T) {
...
+		if err := manager.PutABI(address, abi); err != nil {
+			t.Error(err)
+		}
+		abi2, err := manager.GetABI(address)
+		if err != nil {
+			if db.IsNotFound(err) {
+				t.Errorf("ABI not found for address %s", address)
+			} else {
+				t.Error(err)
+			}
```
Comment (tshakalekholoane) [inline internal/db/abi/manager_test.go:30]: `else` clause can be omitted e.g.

```go
  ...
  t.Errorf("ABI not found for address %s", address)
}
t.Error(err)
```

File: internal/db/block/block_test.go [hunk 7]
```diff
@@ -41,18 +41,29 @@ func TestManager(t *testing.T) {
...
-		returnedBlock := manager.GetBlockByHash(key)
-		if returnedBlock == nil {
-			t.Errorf("unexpected nil after search for block with hash %s", block.BlockHash)
+		returnedBlock, err := manager.GetBlockByHash(key)
+		if err != nil {
+			if db.IsNotFound(err) {
+				t.Errorf("unexpected nil after search for block with hash %s", block.BlockHash)
+			} else {
+				t.Error(err)
+			}
```
Comment (tshakalekholoane) [inline internal/db/block/block_test.go:54]: `else` clause can be omitted. See comment above for an example.

File: internal/db/state/code_test.go [hunk 8]
```diff
@@ -43,8 +43,17 @@ func TestManager_Code(t *testing.T) {
...
+		if err := manager.PutCode(code.Address, code.Code); err != nil {
+			t.Error(err)
+		}
+		obtainedCode, err := manager.GetCode(code.Address)
+		if err != nil {
+			if db.IsNotFound(err) {
+				t.Errorf("Code not found for address %s", code.Address)
+			} else {
+				t.Error(err)
+			}
```
Comment (tshakalekholoane) [inline internal/db/state/code_test.go:55]: `else` clause can be omitted. See comment above for an example.

File: internal/db/transaction/manager_test.go [hunk 9]
```diff
@@ -96,11 +98,20 @@ func TestManager_GetTransaction(t *testing.T) {
...
 	// Get all the transactions and compare
 	for _, tx := range txs {
-		outTx := manager.GetTransaction(tx.GetHash())
+		outTx, err := manager.GetTransaction(tx.GetHash())
+		if err != nil {
+			if db.IsNotFound(err) {
+				t.Errorf("Transaction %s not found", tx.GetHash())
+			} else {
+				t.Error(err)
+			}
```
Comment (tshakalekholoane) [inline internal/db/transaction/manager_test.go:113]: `else` clause can be omitted. See comment above for an example.

File: pkg/rpc/api.go [hunk 10]
```diff
@@ -339,13 +368,25 @@ func (HandlerRPC) StarknetGetTransactionByBlockHashAndIndex(c context.Context, b
...
 		}
 		txHash := block.TxHashes[index]
-		txn := services.TransactionService.GetTransaction(txHash)
+		txn, err := services.TransactionService.GetTransaction(txHash)
+		if db.IsNotFound(err) {
+			// notest
+			return nil, fmt.Errorf("transaction not found")
+		} else if err != nil {
+			return nil, fmt.Errorf("database error: %v", err)
+		}
```
Comment (tshakalekholoane) [inline pkg/rpc/api.go:389]: From the comment above.

> It seems like we might want to check first whether an error occured before asserting its kind and not the other way around just as you did above. I also don't see why the assertion has to be made if there is not way for the client to know which kind of error occured outside of say parsing the first few characters of the error string.

File: pkg/rpc/api.go [hunk 11]
```diff
@@ -414,11 +455,16 @@ func (HandlerRPC) StarknetGetCode(
...
-	code := services.StateService.GetCode(contractAddress.Bytes())
-	if code == nil {
+	code, err := services.StateService.GetCode(contractAddress.Bytes())
+	if db.IsNotFound(err) {
 		// notest
 		return nil, fmt.Errorf("code not found")
+	} else if err != nil {
+		// notest
+		return nil, fmt.Errorf("database error: %v", err)
 	}
```
Comment (tshakalekholoane) [inline pkg/rpc/api.go:468]: From the comment above.

> It seems like we might want to check first whether an error occured before asserting its kind and not the other way around just as you did above. I also don't see why the assertion has to be made if there is not way for the client to know which kind of error occured outside of say parsing the first few characters of the error string.

File: internal/services/transactions_test.go [hunk 12]
```diff
@@ -108,11 +110,20 @@ func TestManager_GetTransaction(t *testing.T) {
...
 	// Get all the transactions and compare
 	for _, tx := range txs {
-		outTx := TransactionService.GetTransaction(tx.GetHash())
+		outTx, err := TransactionService.GetTransaction(tx.GetHash())
+		if err != nil {
+			if db.IsNotFound(err) {
+				t.Errorf("transaction not found: %s", tx.GetHash().String())
+			} else {
+				t.Error(err)
+			}
```
Comment (tshakalekholoane) [inline internal/services/transactions_test.go:125]: `else` clause can be omitted. See comment above for an example.

**Review by tshakalekholoane — CHANGES_REQUESTED** [review]

File: internal/db/abi/manager_test.go [hunk 13]
```diff
@@ -18,10 +18,16 @@ func TestManager(t *testing.T) {
...
-		manager.PutABI(address, abi)
-		abi2 := manager.GetABI(address)
-		if abi2 == nil {
-			t.Errorf("ABI with key %s not found after insertion with the same key", address)
+		if err := manager.PutABI(address, abi); err != nil {
+			t.Error(err)
+		}
+		abi2, err := manager.GetABI(address)
+		if err != nil {
+			if db.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (stdevAdrianPaez) [reply internal/db/abi/manager_test.go:26]: There is no difference, I can remove the `db.IsNotFound` and use `errors.Is` directly, which can improve the code readability

**Review by tshakalekholoane — CHANGES_REQUESTED** [review]
See [unresolved conversation](https://github.com/NethermindEth/juno/pull/268#discussion_r913839812).

File: internal/db/block/manager.go (hunk 3 above)
Comment (adriantpaez) [reply internal/db/block/manager.go:75]: In this case I need the `rawValue` below and I don't want to use an else statement there

**Review by joshklop — APPROVED** [review]

**Review by tshakalekholoane — APPROVED** [review]

**Pull request merged**